### ISSUE TITLE
AttributeError: 'CTClassifier' object has no attribute 'estimator1'

### ISSUE THREAD

**Issue by alysiayx** [mvlearn/mvlearn#317]
Hi everyone,

the CTClassifier failed to run as no attribute 'estimator1' was found. Can anyone fix this issue?

**Comment by agramfort** [comment]
what is the code you used?

**Comment by alysiayx** [comment]
> what is the code you used?
> […](#)
> Message ID: ***@***.***>

I used the example you provided on https://mvlearn.github.io/references/semi_supervised.html?highlight=ctclassifier#mvlearn.semi_supervised.CTClassifier

**Comment by gavinmischler** [comment]
Can you give the details of your environment? I just tried running in a fresh environment with python 3.9 and mvlearn 0.5.0 installed from conda-forge and I was able to run that example code:

```python
from mvlearn.semi_supervised import CTClassifier
from mvlearn.datasets import load_UCImultifeature
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split
data, labels = load_UCImultifeature(select_labeled=[0,1])
X1 = data[0]  # Only using the first view
X1_train, X1_test, l_train, l_test = train_test_split(X1, labels)

estimator1 = GaussianNB()
estimator2 = RandomForestClassifier()
ctc = CTClassifier(estimator1, estimator2, random_state=1)
ctc = ctc.fit([X1_train, X1_train], l_train)
preds = ctc.predict([X1_test, X1_test])
print("Accuracy: ", sum(preds==l_test) / len(preds))
```

**Comment by alysiayx** [comment]
> Can you give the details of your environment? I just tried running in a fresh environment with python 3.9 and mvlearn 0.5.0 installed from conda-forge and I was able to run that example code:
> 
> ```python
> from mvlearn.semi_supervised import CTClassifier
> from mvlearn.datasets import load_UCImultifeature
> import numpy as np
> from sklearn.ensemble import RandomForestClassifier
> from sklearn.naive_bayes import GaussianNB
> from sklearn.model_selection import train_test_split
> data, labels = load_UCImultifeature(select_labeled=[0,1])
> X1 = data[0]  # Only using the first view
> X1_train, X1_test, l_train, l_test = train_test_split(X1, labels)
> 
> estimator1 = GaussianNB()
> estimator2 = RandomForestClassifier()
> ctc = CTClassifier(estimator1, estimator2, random_state=1)
> ctc = ctc.fit([X1_train, X1_train], l_train)
> preds = ctc.predict([X1_test, X1_test])
> print("Accuracy: ", sum(preds==l_test) / len(preds))
> ```


Thank you for your reply. Yes the example can be run as I also use the latest mvlearn, but an error is raised when I try to look at the ctc "AttributeError: 'CTClassifier' object has no attribute 'estimator1'"

I need to tune the hyparameters of the estimator, so this error stops the tuning process.

Indeed the CTClassifier has no attribute "estimator1", instead it has "estimator1_"

**Comment by gavinmischler** [comment]
Right, the estimator1 can be accessed as ``ctc.estimator1_``, but that is the same estimator object that the class was initialized with so you should be able to tune the hyper parameters of each estimator by accessing it with the underscore name.

If you want to just edit the hyperparameters before initializing the CTClassifier, you can just initialize the estimator with different kwargs before passing it in. For example:

```python
estimator1 = GaussianNB(var_smoothing=1e-7)
estimator2 = GaussianNB(var_smoothing=1e-5)
ctc_1 = CTClassifier(estimator1, estimator2, random_state=1)

estimator1 = GaussianNB(var_smoothing=1e-2)
estimator2 = GaussianNB(var_smoothing=1e-5)
ctc_1 = CTClassifier(estimator1, estimator2, random_state=1)
```

**Comment by alysiayx** [comment]
> Right, the estimator1 can be accessed as `ctc.estimator1_`, but that is the same estimator object that the class was initialized with so you should be able to tune the hyper parameters of each estimator by accessing it with the underscore name.
> 
> If you want to just edit the hyperparameters before initializing the CTClassifier, you can just initialize the estimator with different kwargs before passing it in. For example:
> 
> ```python
> estimator1 = GaussianNB(var_smoothing=1e-7)
> estimator2 = GaussianNB(var_smoothing=1e-5)
> ctc_1 = CTClassifier(estimator1, estimator2, random_state=1)
> 
> estimator1 = GaussianNB(var_smoothing=1e-2)
> estimator2 = GaussianNB(var_smoothing=1e-5)
> ctc_1 = CTClassifier(estimator1, estimator2, random_state=1)
> ```

Thank you so much. But if I wanna using GridSearchCV or other similar tools to tune the hyparemeters, I need to define a search space like:

`search_space = {
    'estimator1__var_smoothing': Real(1e-10, 1e-1, "log-uniform"),
  }
`
But it does not work out if the attribute name already has an underscore. Do you know how to solve that?

and also it is strange that the ctc instance cannot be accessed (an error is raised), like for a simple SVM example. the output will be `SVR()` :

```
from sklearn import svm

X = [[0, 0], [2, 2]]
y = [0.5, 2.5]
regr = svm.SVR()
regr.fit(X, y)
regr.predict([[1, 1]])
print(regr)
```

**Comment by gavinmischler** [comment]
Oh thanks for the explanation, now I understand the use-case. I think you're right that this is a bug. I believe (@agramfort please correct me if I'm mistaken) sklearn's BaseEstimator expects parameters passed in an __init__() to not have trailing underscores when they get set as attributes. But, CTClassifier and CTRegressor have init params which get set as self.param_ instead of self.param and this is also what messes up the printing of the class itself, as well as referencing them by GridSearchCV since the get_params() function doesn't work because of this.

It would be a simple fix to change how these attributes are named in the class
